Allow window to remember its size, position and state (GTK + Avalonia)

MutantAura opened the pull request:
Based on #4439 with support for Avalonia. 

As the two frameworks fight over data types and the like, GTK is given priority for now. It will be a quick job to edit and expand upon once deprecated e.g. WindowState could start in fullscreen instead of just maximised etc.

CC @HaizenTrist

gdkchan commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+            ConfigurationState.Instance.Ui.WindowPositionX.Value = window.Position.X;
+            ConfigurationState.Instance.Ui.WindowPositionY.Value = window.Position.Y;
+
+            ConfigurationState.Instance.Ui.WindowMaximized.Value = window.WindowState is WindowState.Maximized ? true : false;
```
This looks very weird, why can't it just do `window.WindowState == WindowState.Maximized` directly?

MutantAura replied:
Not sure why I did that... changed.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -1293,6 +1298,21 @@ private void Exit_Pressed(object sender, EventArgs args)
         {
             if (!_gameLoaded || !ConfigurationState.Instance.ShowConfirmExit || GtkDialog.CreateExitDialog())
             {
+                int windowWidth;
```
Should be probably be moved to a separate method instead of repeating this code twice?

MutantAura replied:
Moved.

AcK77 commented on Ryujinx.Ui.Common/Configuration/ConfigurationFileFormat.cs:
```diff
@@ -300,6 +300,31 @@ public class ConfigurationFileFormat
+        /// <summary>
+        /// Width of the main window in pixels. 
+        /// </summary>
+        public int WindowSizeWidth { get; set;}
```
It’s not possible to use `Rectangle` type instead? Since it already have X, Y, Width and Height field. Maybe the JSON serializing will cause issue.

MutantAura replied:
This would involve introducing a System.Drawing dep in a few places and I'm not entirely sure how the JSON would parse it. I could give it a try but honestly not sure if the result would be any better.

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -62,6 +62,8 @@ public MainWindow()
 
             DataContext = ViewModel;
 
+            SetWindowSizePosition(this);
```
```suggestion
            SetWindowSizePosition();
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
             LoadHotKeys();
         }
 
+        private void SetWindowSizePosition(MainWindow window)
```
```suggestion
        private void SetWindowSizePosition()
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+
+            ViewModel.WindowState = ConfigurationState.Instance.Ui.WindowMaximized.Value is true ? WindowState.Maximized : WindowState.Normal;
+
+            window.Position = new PixelPoint(ConfigurationState.Instance.Ui.WindowPositionX, 
```
```suggestion
            Position = new PixelPoint(ConfigurationState.Instance.Ui.WindowPositionX, 
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+                                             ConfigurationState.Instance.Ui.WindowPositionY);
+        }
+
+        private void SaveWindowSizePosition(MainWindow window)
```
```suggestion
        private void SaveWindowSizePosition()
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+
+        private void SaveWindowSizePosition(MainWindow window)
+        {
+            // Avalonia/GTK use different structures. Conversion is required for inter-operation.
```
```
            // Avalonia/GTK use different structures. Conversion is required for inter-operation.
```
It doesn’t make sense to have a comment about GTK in Avalonia, because when we will drop GTK, it will be pointless

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+            ConfigurationState.Instance.Ui.WindowSizeHeight.Value = (int)window.Height;
+            ConfigurationState.Instance.Ui.WindowSizeWidth.Value = (int)window.Width;
+
+            ConfigurationState.Instance.Ui.WindowPositionX.Value = window.Position.X;
```
```suggestion
            ConfigurationState.Instance.Ui.WindowPositionX.Value = Position.X;
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+            ConfigurationState.Instance.Ui.WindowPositionX.Value = window.Position.X;
+            ConfigurationState.Instance.Ui.WindowPositionY.Value = window.Position.Y;
+
+            ConfigurationState.Instance.Ui.WindowMaximized.Value = window.WindowState == WindowState.Maximized;
```
```suggestion
            ConfigurationState.Instance.Ui.WindowMaximized.Value = WindowState == WindowState.Maximized;
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,31 @@ private void Load()
+            ConfigurationState.Instance.Ui.WindowSizeWidth.Value = (int)window.Width;
+
+            ConfigurationState.Instance.Ui.WindowPositionX.Value = window.Position.X;
+            ConfigurationState.Instance.Ui.WindowPositionY.Value = window.Position.Y;
```
```suggestion
            ConfigurationState.Instance.Ui.WindowPositionY.Value = Position.Y;
```

AcK77 commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -388,6 +415,8 @@ protected override void OnClosing(CancelEventArgs e)
                 return;
             }
 
+            SaveWindowSizePosition(this);
```
```suggestion
            SaveWindowSizePosition();
```

AcK77 commented on Ryujinx.Ui.Common/Configuration/ConfigurationState.cs:
```diff
@@ -713,6 +748,9 @@ public void LoadDefault()
             Ui.IsAscendingOrder.Value                 = true;
             Ui.StartFullscreen.Value                  = false;
             Ui.ShowConsole.Value                      = true;
+            Ui.WindowSizeWidth.Value                  = 1280;
```
Could be nice to set PositionX and PositionY to 0 for consistency (even if it’s already 0 as default)

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -152,8 +152,13 @@ private MainWindow(Builder builder) : base(builder.GetRawOwnedObject("_mainWin")
-            DefaultHeight = monitorHeight < 760  ? monitorHeight : 760;
+            DefaultWidth  = monitorWidth  < 1280 ? monitorWidth  : ConfigurationState.Instance.Ui.WindowSizeWidth;
+            DefaultHeight = monitorHeight < 760  ? monitorHeight : ConfigurationState.Instance.Ui.WindowSizeHeight;
+            Move(ConfigurationState.Instance.Ui.WindowPositionX, ConfigurationState.Instance.Ui.WindowPositionY);
```
Add a new empty line before and after

TSRBerry commented:
The network interface dropdown PR also changed the config file version, so you might need to increase it here again

MutantAura commented:
Updated config file, added checks on the Avalonia side to ensure the window opens at a valid screen co-ordinate. Previously you could feed it arbitrary values/adjust your monitor config to break it.

TSRBerry commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -1330,6 +1339,25 @@ private void Window_Close(object sender, DeleteEventArgs args)
 
+        private void SaveWindowSizePosition()
+        {
+            int windowWidth;
+            int windowHeight;
+            int windowXPos;
+            int windowYPos;
+
+            GetSize(out windowWidth, out windowHeight);
+            GetPosition(out windowXPos, out windowYPos);
```
```suggestion
            GetSize(out int windowWidth, out int windowHeight);
            GetPosition(out int windowXPos, out int windowYPos);
```
Could you inline these?

MutantAura replied:
Yes!

TSRBerry commented on Ryujinx.Ui.Common/Configuration/ConfigurationState.cs:
```diff
@@ -132,6 +132,31 @@ public ShownFileTypeSettings()
             /// Hide / Show Console Window
             /// </summary>
             public ReactiveObject<bool> ShowConsole { get; private set; }
+            
+            /// <summary>
+            /// Width of the main window in pixels. 
+            /// </summary>
+            public ReactiveObject<int> WindowSizeWidth { get; private set; }
+
+            /// <summary>
+            /// Height of the main window in pixels. 
+            /// </summary>
+            public ReactiveObject<int> WindowSizeHeight { get; private set; }
+
+            /// <summary>
+            /// Horizontal position of the main window in pixels. 
+            /// </summary>
+            public ReactiveObject<int> WindowPositionX { get; private set; }
+
+            /// <summary>
+            /// Vertical position of the main window in pixels. 
+            /// </summary>
+            public ReactiveObject<int> WindowPositionY { get; private set; }
+            
+            /// <summary>
+            /// Whether the main window is maximized or not. 
+            /// </summary>
+            public ReactiveObject<bool> WindowMaximized { get; private set; }
```
Could you group these inside a `Window` subclass like the settings in `ShownFileTypeSettings` are?

MutantAura replied:
Also yes

TSRBerry reviewed (commented):
Lgtm! I Just have 2 styling related comments.

gdkchan commented on Ryujinx.Ui.Common/Configuration/ConfigurationFileFormat.cs:
```diff
@@ -305,6 +305,31 @@ public class ConfigurationFileFormat
+        /// <summary>
+        /// Width of the main window in pixels. 
+        /// </summary>
+        public int WindowSizeWidth { get; set;}
```
@TSRBerry Would the formatting check be able to catch this? (I did not check if whitespace formatting is enabled, I think it was disabled).

TSRBerry replied:
It's currently disabled, because it would remove all of the vertical alignments we have.

I was thinking a bit about this, but was also hoping for some input on #4670 regarding this.

MutantAura replied:
Should all have correct spacing now.

TSRBerry reviewed (approved):
Awesome, thank you!

marysaka commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml:
```diff
@@ -12,15 +12,14 @@
     WindowState="{Binding WindowState}"
-    Width="1280"
-    Height="777"
+    Width="{Binding WindowWidth}"
+    Height="{Binding WindowHeight}"
```
I don't see the default value being set on MainWindowViewModel, is that intended?

I think it's fine but that could maybe mess around with the UI designer. (if it even work)

MutantAura replied:
I can set a default but it should fall back to MinWidth/MinHeight in these scenarios, in all other situations it should pull from the config file (via migration or generation).

marysaka commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,46 @@ private void Load()
+
+            ViewModel.WindowState = ConfigurationState.Instance.Ui.WindowStartup.WindowMaximized.Value is true ? WindowState.Maximized : WindowState.Normal;
+        
+            Position = CheckScreenBounds(SavedPoint) ? SavedPoint : new PixelPoint(0,0);
```
Is (0,0) the center of the screen? If not that change the behavior we currently have. (aka centered)

MutantAura replied:
Fixed.

marysaka commented on Ryujinx.Ui.Common/Configuration/ConfigurationFileFormat.cs:
```diff
@@ -251,6 +251,11 @@ public class ConfigurationFileFormat
         public ShownFileTypes ShownFileTypes { get; set; }
 
+        /// <summary>
+        /// A list of file types to be hidden in the games List
```
wrong comment copy pasta.
```suggestion
        /// Main window start-up position, size and state
```

MutantAura replied:
Fixed.

marysaka commented on Ryujinx.Ui.Common/Configuration/ConfigurationState.cs:
```diff
@@ -1328,19 +1368,33 @@ public ConfigurationLoadResult Load(ConfigurationFileFormat configurationFileFor
                     NRO  = true,
                     NSO  = true
                 };
-
-                configurationFileUpdated = true;
```
This shouldn't have been removed.

MutantAura replied:
Fixed.

marysaka commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -158,8 +158,15 @@ private MainWindow(Builder builder) : base(builder.GetRawOwnedObject("_mainWin")
 
-            DefaultWidth  = monitorWidth  < 1280 ? monitorWidth  : 1280;
-            DefaultHeight = monitorHeight < 760  ? monitorHeight : 760;
+            DefaultWidth  = monitorWidth  < 1280 ? monitorWidth  : ConfigurationState.Instance.Ui.WindowStartup.WindowSizeWidth;
+            DefaultHeight = monitorHeight < 760  ? monitorHeight : ConfigurationState.Instance.Ui.WindowStartup.WindowSizeHeight;
```
This logic differ from Avalonia here.
I do think we should have the same monitor size check in Avalonia too.

marysaka replied:
For this I think it's fine for now but we should probably keep the same logic between the two (aka remove this check or reproduce it if possible)

MutantAura replied:
As outlined on discord I think it's best to remove. The theoretical situation this "solves" is virtually impossible and potentially undesirable. I'll try and speed up the work on: https://github.com/Ryujinx/Ryujinx/pull/4681 as I think it's a better solution if this were to ever be a problem.

marysaka reviewed (commented):
Thanks for the PR, I do have some comments

gdkchan commented on Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:
```diff
@@ -297,6 +299,46 @@ private void Load()
+                }
+            }
+
+            Logger.Warning?.Print(LogClass.Application, $"Failed to find valid start-up coordinates. Defaulting to 0,0.");
```
```suggestion
            Logger.Warning?.Print(LogClass.Application, $"Failed to find valid start-up coordinates. Defaulting to 0, 0.");
```

marysaka reviewed (approved):
Need a rebase but lgtm otherwise :+1:

MutantAura commented:
Rebased.
